app/models/pillar.rb

```diff
@@ -54,5 +54,22 @@ def all_pillars
         apiserver:    "api:server:external_fqdn"
       }
     end
+
+    # Apply the given pillars into the database. It returns an array with the
+    # encountered errors.
+    def apply(pillars)
+      errors = []
+
+      Pillar.all_pillars.each do |key, pillar_key|
+        pillar = Pillar.find_or_initialize_by pillar: pillar_key
+        pillar.value = pillars[key]
+        next if pillar.save
+
+        exp = pillar.errors.messages.empty? ? "" : ": #{pillar.errors.messages[:value].first}"
```

Comment:
In the near future we can even add `pillar.errors.add(pillars[key], ...` to add errors to their specific virtual pillar attribute, and we could highlight specific fields on the forms, if required.